src/fluree/db/json_ld/policy.cljc

```diff
@@ -497,22 +498,28 @@
   (-> (select-keys opts [:did :role :credential])
       not-empty))
 
-(defn roles-for-did
-  [db did]
-  (go-try
-    (try
-      (->> (<? (fql/query db {"select" "?role"
-                              "where" [[did "https://ns.flur.ee/ledger#role" "?role"]]}))
-           (not-empty))
-      ;; in case no "f:role" exists
-      (catch Exception _))))
+(defn roles-for-sid
+  [db sid]
+  (let [in-ch        (async/chan 2 cat)
+        out-ch       (async/chan 2 (comp cat (map flake/o)))
+        did-roles-ch (query-range/index-range db :spot = [sid const/$_role])]
```

Comment:
I didn't realize during my last review that roles were operating on iris and not sids. I don't know if I would have made the suggestion had I realized that because of the extra hassle of looking up all those iris, but using `index-range` is still much faster than running an entire query, so I'll leave it to you to decide which method you go with. 

If you do go the `index-range` route, then I have more suggestions. There is a (sadly undocumented) option to `index-range` that allows you to pass a transducer to transform all the flakes directly, so you don't have to create an extra channel. You can pass an options map using [this arity of the `index-range` function](https://github.com/fluree/db/blob/main/src/fluree/db/query/range.cljc#L319-L321), and you can specify a transducer to transform all the flakes with the `:flake-xf` key.

Also, `index-range` is different from `tree-chan` in that it is eager and will always return the entire collection of flakes, and it could possibly return an exception if there was _any_ error at any point in its processing, so this is one of those functions that return channels that act like promises and should be accessed with `<?`. `pipeline-async` won't work without explicit error handling with an error channel, which will introduce even more complexity.

Instead of introducing error channels and all that, you could do something like this:

```clojure
(defn roles-for-sid
  [db sid]
  (go-try
    (let [did-role-sids (<? (query-range/index-range db :spot = [sid const/$_role]
                                                     {:flake-xf (map flake/o)}))]
      (loop [[did-role-sid & r] did-role-sids
             did-role-iris      []]
        (if did-role-sid
          (let [did-role-iri (<? (query-range/index-range db :spot = [did-role-sid const/$xsd:anyURI]))]
            (recur r (conj did-role-iris did-role-iri)))
          did-role-iris)))))
```